File: go/transfer.go
package main

import (
	"database/sql"
	"encoding/json"
	"flag"
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"time"

	_ "github.com/go-sql-driver/mysql"
	_ "github.com/lib/pq"
)

// MySQL timestamp format.
const datetimeLayout = "2006-01-02 15:04:05"

// Print or not to print SQL queries.
var isDebug = false

// The config file.
type config struct {
	SourceDSN string `json:"sourceDB"`
	TargetDSN string `json:"targetDB"`
}

// To the target DB.
type event struct {
	last   time.Time
	metric string
	count  int
}

// logDebug print the text to the log output if isDebug is true.
func logDebug(text string) {
	if isDebug {
		log.Println(text)
	}
}

// parseConfig reads and parse the config JSON file.
func parseConfig() (conf config, err error) {
	content, err := ioutil.ReadFile("config.json")
	if err != nil {
		return
	}

	err = json.Unmarshal(content, &conf)
	return
}

// isTableExists checks is the table in the DB or not.
func isTableExists(db *sql.DB, tableName string) bool {
	query := fmt.Sprintf("SELECT 1 FROM %s LIMIT 1;", tableName)
	logDebug(query)

	_, err := db.Exec(query)
	if err != nil {
		return false
	}

	return true
}

// createTablesIfNotExist checks the tables in the provided DB and creates them if aren't there.
func createTablesIfNotExist(db *sql.DB) (err error) {
	if !isTableExists(db, "metrics") {
		query := `CREATE TABLE metrics(
			id SERIAL PRIMARY KEY NOT null,
			name VARCHAR(20) NOT null
			);`
		logDebug(query)

		_, err = db.Exec(query)
		if err != nil {
			return err
		}
	}
	if !isTableExists(db, "events") {
		query := `CREATE TABLE events(
			id SERIAL PRIMARY KEY NOT NULL,
			day_last TIMESTAMP NOT NULL,
			count INT NOT NULL,
			metric_id INT,
			FOREIGN KEY(metric_id) REFERENCES metrics(id) ON DELETE CASCADE
			);`
		logDebug(query)

		_, err = db.Exec(query)
		if err != nil {
			return err
		}
	}

	return
}

// getLastMetricsTime query the DB for the last times for each metrics occured.
func getLastMetricsTime(db *sql.DB) (lasts map[string]time.Time, err error) {
	query := `SELECT metrics.name, max(day_last) as last
		FROM events
		INNER JOIN metrics ON events.metric_id = metrics.id
		GROUP BY name;`
	logDebug(query)

	res, err := db.Query(query)
	if err != nil {
		return
	}
	defer res.Close()

	lasts = map[string]time.Time{}

	for res.Next() {
		var name, temp string
		var last time.Time

		err = res.Scan(&name, &temp)
		if err != nil {
			return
		}

		last, err = time.Parse(time.RFC3339, temp)
		if err != nil {
			return nil, err
		}

		lasts[name] = last
	}

	return
}

// getOrCreateMetric returns the ID of a metric by name or create a new one.
func getOrCreateMetric(db *sql.DB, name string) (id int, err error) {
	query := fmt.Sprintf("SELECT id FROM metrics WHERE name = '%s';", name)
	logDebug(query)

	err = db.QueryRow(query).Scan(&id)
	if err == sql.ErrNoRows {
		err = nil

		query = fmt.Sprintf("INSERT INTO metrics (name) VALUES ('%s');", name)
		logDebug(query)

		_, err := db.Exec(query)
		if err != nil {
			return 0, err
		}

		id, err = getOrCreateMetric(db, name)
	}

	return
}

func main() {
	// Get the CLI flag.
	isDebugPtr := flag.Bool("d", false, "log SQL queries to stdout")
	flag.Parse()
	isDebug = *isDebugPtr

	// Parse the config file if exists.
	conf, err := parseConfig()
	if os.IsNotExist(err) {
		log.Println("config.json not found. Continue with default values.")
	} else if err != nil {
		log.Fatal(err)
	}

	// Prepare the data source names.
	var dsnSource, dsnTarget string
	if conf.SourceDSN != "" {
		dsnSource = conf.SourceDSN
	} else {
		dsnSource = "liligo:liligo@tcp(localhost:3306)/liligo"
	}
	if dsnTarget != "" {
		dsnTarget = conf.TargetDSN
	} else {
		dsnTarget = "postgres://liligo:liligo@localhost:5432/liligo?sslmode=disable"
	}

	// Create connection to the databases.
	dbSource, err := sql.Open("mysql", dsnSource)
	if err != nil {
		log.Fatal(err)
	}
	defer dbSource.Close()

	dbTarget, err := sql.Open("postgres", dsnTarget)
	if err != nil {
		log.Fatal(err)
	}
	defer dbTarget.Close()

	// Create tables if needed.
	if err = createTablesIfNotExist(dbTarget); err != nil {
		log.Fatal("Error while creating the tables. ", err)
	}

	// Get the last times for each metric in the events database.
	lasts, err := getLastMetricsTime(dbTarget)
	if err != nil {
		log.Fatal(err)
	}

	// Fetch all the data from the source DB.
	query := "SELECT `index`, `timestamp`, metric FROM liligo"
	logDebug(query)

	res, err := dbSource.Query(query)
	if err != nil {
		log.Fatal(err)
	}
	defer res.Close()

	// Process the records.
	processed := map[string]event{}
	for res.Next() {
		var index, temp, metric string
		var timestamp time.Time

		// Scan the next row.
		err = res.Scan(&index, &temp, &metric)
		if err != nil {
			log.Fatal(err)
		}

		// Parse the datetime string.
		timestamp, err = time.Parse(datetimeLayout, temp)
		if err != nil {
			log.Fatal(err)
		}

		// Check the time of this record. Modify history is NOT allowed.
		last, exists := lasts[metric]
		if exists && last.After(timestamp) || last.Equal(timestamp) {
			continue
		}

		// Contruct a key for the map.
		key := metric + timestamp.String()[:11]

		// Add to the map or update existing element.
		value, exists := processed[key]
		if !exists {
			value = event{
				last:   timestamp,
				metric: metric,
				count:  1,
			}
		} else {
			value.count++
			if value.last.Before(timestamp) {
				value.last = timestamp
			}
		}

		processed[key] = value
	}

	// Insert the processed data to the new DB.
	for _, value := range processed {
		// Get the ID for the metric. If not exitst create one and return that new ID.
		mID, err := getOrCreateMetric(dbTarget, value.metric)
		if err != nil {
			log.Fatal(err)
		}

		// Create the insert command,
		query := fmt.Sprintf("INSERT INTO events (day_last, count, metric_id) VALUES('%s', %d, %d)", value.last.Format(time.RFC3339), value.count, mID)
		logDebug(query)

		// then execute it.
		_, err = dbTarget.Exec(query)
		if err != nil {
			log.Fatal(err)
		}
	}

	fmt.Println("DONE")
}

File: python/transfer.py
#!/usr/bin/env python3

import os
from datetime import datetime
from argparse import ArgumentParser

from sqlalchemy.sql import func
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker, Session, relationship
from sqlalchemy import create_engine, Column, Integer, String, DateTime, ForeignKey


BaseSource = declarative_base()
BaseTarget = declarative_base()

class Liligo(BaseSource):
    __tablename__ = "liligo"

    index = Column(Integer, primary_key=True)
    timestamp = Column(DateTime)
    metric = Column(String)

class Metrics(BaseTarget):
    __tablename__ = "metrics"

    id = Column(Integer, primary_key=True)
    name = Column(String)

class Events(BaseTarget):
    __tablename__ = "events"

    id = Column(Integer, primary_key=True)
    day_last = Column(DateTime)
    count = Column(Integer)
    metric_id = Column(Integer, ForeignKey("metrics.id"))

    metric = relationship("Metrics")

class Data:
    def __init__(self, last: DateTime):
        self._count = 1
        self._last = last

    def increase(self, when: DateTime):
        """Increase the inner counter by 1 and set the new datetime as the last one if it's newer.

        Args:
            last (DateTime): last datetime value
        """

        self._count += 1

        if self._last < when:
            self._last = when

    def get(self) -> tuple:
        """Returns the count and last values.

        Returns:
            tuple: (count, last)
        """

        return (self._count, self._last)

def get_or_create_metric(session: Session, metric: str) -> int:
    """Returns the ID of the given metric. If not exists create it.

    Args:
        session (Session): current DB session
        metric (str): name of the metric

    Returns:
        int: ID of the metric
    """

    mid = session.query(Metrics.id).filter(Metrics.name==metric).first()
    if mid is not None:
        return mid[0]

    session.add(Metrics(name=metric))

    return session.query(Metrics.id).filter(Metrics.name==metric).first()[0]

def main():
    # Parse the CLI arguments.
    parser = ArgumentParser(description="Wanna be an analytics developer.")
    parser.add_argument("-d", "--debug", action="store_true", help="show SQL logs")
    args = parser.parse_args()

    # Get the URLs for the DB connections.
    source_db_url = os.getenv("SOURCE_DB_URL", default="mysql+pymysql://liligo:liligo@localhost:3306/liligo")
    target_db_url = os.getenv("TARGET_DB_URL", default="postgresql://liligo:liligo@localhost:5432/liligo")

    # Add pymysql driver to the MySQL DB URL if necessary.
    source_db_url = source_db_url if "mysql+pymysql://" in source_db_url else source_db_url.replace("mysql://", "mysql+pymysql://")

    # Create the DB engines.
    engine_source = create_engine(source_db_url, echo=args.debug)
    engine_target = create_engine(target_db_url, echo=args.debug)

    # Create the tables in the target DB.
    try:
        BaseTarget.metadata.create_all(engine_target, checkfirst=True)
    except Exception as e:
        print("Cannot create the tables in the target DB.", e)
        return

    # Create sessions.
    SessionSource = sessionmaker(bind=engine_source)
    SessionTarget = sessionmaker(bind=engine_target)
    session_source = SessionSource()
    session_target = SessionTarget()

    # Get last events datetime for each metrics in the target DB.
    res = session_target.query(Metrics.name, func.max(Events.day_last)).filter(Events.metric_id==Metrics.id).group_by(Metrics.name).all()
    current = {}
    for row in res:
        current[row[0]] = row[1]

    # Process the records from the source database.
    # For more CPU intensive tasks we should use multiprocessing with queue.
    processed = {}
    for i in session_source.query(Liligo).yield_per(10000):
        date = i.timestamp.date()

        last = current.get(i.metric)
        if last is not None and last >= i.timestamp:
            continue

        if (i.metric, date) not in processed:
            processed[(i.metric, date)] = Data(i.timestamp)
            continue

        processed[(i.metric, date)].increase(i.timestamp)

    # Create the new records in the target DB.
    new_records = []
    for key, value in processed.items():
        mid = get_or_create_metric(session_target, key[0])

        count, last = value.get()

        new_records.append(
            Events(count=count, day_last=last, metric_id=mid)
        )

    # Save the new modifications to the DB.
    session_target.add_all(new_records)
    session_target.commit()

    # Close the sessions.
    session_source.close()
    session_target.close()

    print("DONE")


if __name__ == "__main__":
    main()
